lib/protocol/bep_extensions.go

```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
 	return f.Version.Counters == nil
 }
 
+// IsEqual checks that the two file infos represent the same actual file.
+// Permissions (config) and blocks (scanning) can be excluded from the comparison.
+// Any file info is not "equal", if it has different
+//  - type
+//  - deleted flag
+//  - invalid flag
+//  - permissions, unless they are ignored
+// A file is not "equal", if it has different
+//  - modification time
+//  - size
+//  - blocks, unless there are no blocks to compare (scanning)
+// A symlink is not "equal", if it has different
+//  - target
+// A directory does not have anything specific to check.
+func (f FileInfo) IsEqual(other FileInfo, ignorePerms bool, noBlocks bool) bool {
+	if f.IsEmpty() || other.IsEmpty() {
+		return f.IsEmpty() == other.IsEmpty()
+	}
+
+	// characteristics common to all types
+	if f.Type != other.Type || f.Deleted != other.Deleted || f.Invalid != other.Invalid || (!(ignorePerms || f.NoPermissions || other.NoPermissions) && !PermsEqual(f.Permissions, other.Permissions)) {
+		return false
+	}
+
+	switch f.Type {
+	case FileInfoTypeFile:
+		return f.Size == other.Size && f.ModTime().Equal(other.ModTime()) && (noBlocks || BlocksEqual(f.Blocks, other.Blocks))
+	case FileInfoTypeSymlink:
+		return f.SymlinkTarget == other.SymlinkTarget
+	case FileInfoTypeDirectory:
+		return true
+	}
+
+	// Should the deprecates symlinks type be handled somehow?
+
+	return false
+}
+
+func PermsEqual(a, b uint32) bool {
+	switch runtime.GOOS {
+	case "windows":
+		// There is only writeable and read only, represented for user, group
+		// and other equally. We only compare against user.
+		return a&0600 == b&0600
+	default:
+		// All bits count
+		return a&0777 == b&0777
+	}
+}
+
+// BlocksEqual returns whether two slices of blocks are exactly the same hash
+// and index pair wise.
+func BlocksEqual(src, tgt []BlockInfo) bool {
```

Comment:
This used to live in `scanner`, then someone refactored that to a private function in `model` for some reason, and now it's in `protocol`. We should stop moving things around...

Reply:
Yeah - I just needed it here and I certainly wont dip into model from protocol (can't anyway because circular) and as I thus had to move it, protocol felt like the right place (it's operating on protocol.BlockInfo after all).